### ISSUE TITLE
[1.20.1] Bug or mod interaction issue with Str buffs and Dagger of Sacrafice causig no LP generation at alter

### ISSUE THREAD

**Issue by Warzer1** [WayofTime/BloodMagic#2052]
#### Issue Description:
Posted here first: as its with ATM9:tts 1.0.7 https://github.com/AllTheMods/All-the-mods-9-Sky/issues/508
Since I think this is an issue with the Sacrificial Dagger's execute I am also reporting it here as well.

I was getting LP in about 1 in 10 dagger kills at the blood alter.  After some digging i found that if i remove the strength buff from the Everlasting abilities mod, I would start getting LP again.  I do think there might be 1 or more other buffs that cause this issue but are lost prone to notice compared to the STR one.

#### What happens:
With a STR buff you rarely get LP at the blood alter with the Sacrificial Dagger

#### What you expected to happen:
To get full LP even with a strength buff.

#### Steps to reproduce:

1. Build blood alter and mob farm above it.
2. Use everlasting abilties Strength mod 1-5 (the more str the more this issue occurs)
3. using sacraficial dagger kill mobs
...
Potential cause (guess): I think the extra str might be causing the sacrifice execute feature to go negative on mob health causing this issue.
____
#### Affected Versions (Do *not* use "latest"):

- BloodMagic: alpha: 1.20.1-3.3.3-45
- Minecraft: 1.20.1
- Forge: 47.2.20
- modpack: all the mods 9 to the sky 1.0.7

**Comment by stellanera98** [comment]
Well, strength gives you melee damage which is applied before the dagger does its thing. There are not going to be any negative values since if the (stored floating point value) health is below 0.5 it simply refuses to work.
Its also worth noting that much like every other variant of sacrifice the dagger works off of the health sacrificed which in this case is the health remaining after being hit. Im not sure this is intended to be this way but I would guess that if not the LivingHurtEvent would have been used instead of the item method called when damaging a mob.

I've been testing this with just bloodmagic and executing a villager with full health without strength gives 1900 lp (20 health - 1 damage = 19, 19 * 100 = 1900, checks out).
Doing the same thing with strength 5 only nets 400 lp since the melee attack dealt 16 damage

If you actually got nothing from the villagers they might have been out of range of the altar when you hit them or already took 3.5+ damage

Unrelated to this you're going to be off better either setting up the ritual to automatically sacrifice mobs (which isnt affected by your strength at all) or by building an incense altar and getting some though palms and healthy levels on a set of living armour instead of combining the worst aspects of both methods

**Comment by Warzer1** [comment]
Thanks for the info above, I didn't know they grant different values of LP based on there life.  I put the spawner where i had a mob farm for they fall quite far and some times die on the fall so i think your right that they might be to low of health when i execute them and not getting LP or is so little with STR that it doesn't show a change in the LP count.  They show from .5 hearts to 5 hearts after the fall.  Will try adjusting the fall height and retest with str modifier and see if that changes anything.  
As for distance from alter i tested it when them standing on it directly.  Normally they are a half slab up so i can still see the contents of the alter.

**Comment by stellanera98** [comment]
Should be in range then, but with a half heart even hitting them without strength would reduce them below the threshold. With 5 full hearts you'd get max 900 (without strength, before runes) but strength 3 or more you kill them before the LP gained is calculated.
For the maximum benefit you want the villagers to fall max 3 blocks and have no strength.

**Comment by Warzer1** [comment]
Yeah after some testing they need to be nearly full health with str 5 to get LP.  without str i can get LP when they are at much lower life.
So question is is that str modifier effect on LP gen expected behavior or a bug?  It does help me either way I just want to pass on the correct info on the ATM9:TTS post as well.

**Comment by stellanera98** [comment]
As I said, its because the damage goes through before the health is checked and it would have been relatively simple to make it so it would either take the max health of the mob or the health before the damage went through, so I dont think its unintentional behavior. I do not know what Way thinks about it though so cant say for sure. He's currently busy with other things so you're probably not getting a definitive answer anytime soon.

**Comment by wrincewind** [comment]
I'm going to say that it should probably take the mob's health before damage goes through, rather than health after damage. (I was going to say max health but considered a weird edge-case with temporary bonus hearts - other similar issues may exist. Also damaging mobs down to half a heart before sacrificing. not sure why anyone **would** , but there you go.)

**Comment by stellanera98** [comment]
If dagger activation can be changed to be on right click instead of left click that would be incredibly simple. If not its likely much more complicated. The dagger gets the hp of the mob on (afaik) the first method called for items on left clicking mobs but its still after damage is done. I know the damage events have been reworked for 1.21 soI dont know if there is one that has the pre-damage hp amount, or current hp and damage dealt to math it out

**Comment by wrincewind** [comment]
would just checking for Max Health be easier, then?  or maybe the greater of current health (after attack) and max health?

**Comment by stellanera98** [comment]
That would be an option but you'd get the max health value in every realistic scenario. Maybe adding the users current melee damage to the mobs hp and using that for the conversion? (capped at current max hp of course)

**Comment by wrincewind** [comment]
i was mostly thinking of cases where the mob has an unexpected buff - could see someone setting up something like create's potion dispenser or rftool's effects applicator to give all mobs a buff to their max hp so they're worth more to sacrifice. niche, but possible. 

getting user's damage + mob health sounds reasonable, yeah. I don't think it should be capped at max HP though, in case of the above setup - i'd be OK with giving people a little more LP than they expect, given we've been giving them a little less for years :p
